pkg/kubectl/rolling_updater.go

```diff
@@ -93,27 +86,450 @@ const (
+
+	// Scale newRc and oldRc until newRc has the desired number of replicas and
+	// oldRc has 0 replicas.
+	increment := maxIncrement
+	progressDeadline := time.Now().UnixNano() + config.Timeout.Nanoseconds()
+	for newRc.Spec.Replicas != desired || oldRc.Spec.Replicas != 0 {
+		// Store the existing replica counts for progress timeout tracking.
+		newReplicas := newRc.Spec.Replicas
+		oldReplicas := oldRc.Spec.Replicas
+
+		if scaleUpFirst {
+			// Scale up/down. Initially this means scale up by the max increment.
+			scaledRc, err := r.scaleUp(newRc, oldRc, desired, increment, config)
+			if err != nil {
+				return err
+			}
+			newRc = scaledRc
+			scaleUpFirst = false
+		} // Otherwise, scale down/up.
+
+		// Scale down as much as possible while maintaining the minimum available
+		// amount. The scale-down informs us as to how much we can safely scale up
+		// next interval.
+		scaledRc, newIncrement, err := r.scaleDown(newRc, oldRc, desired, maxIncrement, minAvailable, config)
+		if err != nil {
+			return err
+		}
+		oldRc = scaledRc
+		increment = newIncrement
+
+		// Wait between down/up.
+		time.Sleep(config.UpdatePeriod)
+
+		// Scale up as much as possible.
+		scaledRc, err = r.scaleUp(newRc, oldRc, desired, increment, config)
+		if err != nil {
+			return err
+		}
+		newRc = scaledRc
+
+		// If we are making progress, continue to advance the progress deadline.
+		// Otherwise, time out with an error.
+		progressMade := (newRc.Spec.Replicas != newReplicas) || (oldRc.Spec.Replicas != oldReplicas)
+		if progressMade {
+			progressDeadline = time.Now().UnixNano() + config.Timeout.Nanoseconds()
+		} else if time.Now().UnixNano() > progressDeadline {
+			return fmt.Errorf("timed out waiting for any update progress to be made")
+		}
+	}
+
+	// Housekeeping and cleanup policy execution.
+	return r.cleanup(oldRc, newRc, config)
+}
+
+// scaleUp scales up newRc to desired by increment. It will safely no-op as
+// necessary when it detects redundancy or other relevant conditions.
+func (r *RollingUpdater) scaleUp(newRc, oldRc *api.ReplicationController, desired, increment int, config *RollingUpdaterConfig) (*api.ReplicationController, error) {
+	// If we're already at the desired, do nothing.
+	if newRc.Spec.Replicas == desired {
+		return newRc, nil
+	}
+	// If the current safe increment is 0, we don't have capacity to scale up
+	// while maintaining in-place semantics, so do nothing.
+	if increment == 0 {
+		return newRc, nil
+	}
+	// If the old is already scaled down, go ahead and scale all the way up.
+	if oldRc.Spec.Replicas == 0 {
+		increment = desired - newRc.Spec.Replicas
+	}
+	// Increase the replica count, and deal with fenceposts.
+	newRc.Spec.Replicas += increment
+	if newRc.Spec.Replicas > desired {
+		newRc.Spec.Replicas = desired
+	}
+	// Perform the scale-up.
+	fmt.Fprintf(config.Out, "Scaling %s up to %d\n", newRc.Name, newRc.Spec.Replicas)
+	retryWait := &RetryParams{config.Interval, config.Timeout}
```

Comment:
Use [this](https://github.com/GoogleCloudPlatform/kubernetes/blob/cb49b195a6c2919f76b98b7c8a38d25bc48e4682/pkg/kubectl/scale.go#L109)